### PR TITLE
Fix race condition

### PR DESCRIPTION
**Pull request by j-rf**
Issue - two running process on the same user could copy their pex to the same place
Fix - put operation is slow, but remove and rename/mv are fast:
Since the expected behavior is to overwrite the destination with a newer pex, and move cannot overwrite:

- only remove if we force upload (so that we can do the final move)

- do a put to a temp destination

- rename/mv from temp to final destination

Why we do not overwrite after the put at rename/mv time if the destination exists (race condition is met):

- it would recreate a possibly inconsistent state

- we fail only if the destination is different from the file being pushed to ensure a consistent state

### CODE REVIEW

File: cluster_pack/filesystem.py [hunk 1]
```diff
@@ -203,6 +230,19 @@ def put(self, filename: str, path: str, chunk: int = 2**16) -> None:
                     target.write(out)
         _preserve_acls(self.base_fs, filename, path)
 
+    def put_atomic(self,
+         source_path: str,
+         destination_path: str) -> None:
+        upload_dir = f"{destination_path}.{uuid.uuid4()}.tmp"
```
Comment (Totoketchup) [inline cluster_pack/filesystem.py:236]: [OPT] it's more a temp file here

File: cluster_pack/filesystem.py [hunk 2]
```diff
@@ -203,6 +230,19 @@ def put(self, filename: str, path: str, chunk: int = 2**16) -> None:
                     target.write(out)
         _preserve_acls(self.base_fs, filename, path)
 
+    def put_atomic(self,
+         source_path: str,
+         destination_path: str) -> None:
+        upload_dir = f"{destination_path}.{uuid.uuid4()}.tmp"
+        try:
+            self.put(source_path, upload_dir)
```
Comment (Totoketchup) [inline cluster_pack/filesystem.py:238]: [OPT] This operation will not be the one failing since we are writing on a random temporary location, only the mv op can have a concurrency issue.

File: tests/integration/test_skein_launcher.py [hunk 3]
```diff
@@ -17,8 +17,9 @@
```
Comment (Totoketchup) [inline tests/integration/test_skein_launcher.py:17]: Are these changes related to this issue here ?

File: cluster_pack/uploader.py [hunk 4]
```diff
@@ -309,11 +340,29 @@ def _upload_env_from_venv(
...
 
+def _check_pex_version(resolved_fs: filesystem.EnhancedFileSystem,
+                       local_package_path: str,
+                       remote_package_path: str) -> bool:
+    with tempfile.TemporaryDirectory() as tempdir:
+        local_copy_path = os.path.join(tempdir, os.path.basename(local_package_path))
+        resolved_fs.get(remote_package_path, local_copy_path)
+        into_at_destination = PexInfo.from_pex(local_copy_path)
+        info_to_upload = PexInfo.from_pex(local_package_path)
+    return info_to_upload.code_hash == into_at_destination.code_hash
```
Comment (Totoketchup) [inline cluster_pack/uploader.py:363]: It would have been nicer to extract such changes that are not part of the ticket in another PR but LGTM

File: cluster_pack/filesystem.py [hunk 5]
```diff
@@ -191,6 +211,13 @@ def __init__(self, base_fs: Any):
         if isinstance(base_fs, pyarrow.filesystem.LocalFileSystem):
             base_fs.chmod = types.MethodType(_chmod, base_fs)
             base_fs.rm = types.MethodType(_rm, base_fs)
+            base_fs.rename = types.MethodType(_rename, base_fs)
+
+        if isinstance(base_fs, pyarrow.filesystem.S3FSWrapper):
+            base_fs.rename = types.MethodType(_s3_rename, base_fs)
```
Comment (Totoketchup) [inline cluster_pack/filesystem.py:217]: I don't fully get why we are adding a 'rename' function here, it is not used after. 🤔 
Isn't there any confusion with the `mv` function maybe ?

File: cluster_pack/filesystem.py [hunk 6]
```diff
@@ -203,6 +230,19 @@ def put(self, filename: str, path: str, chunk: int = 2**16) -> None:
                     target.write(out)
         _preserve_acls(self.base_fs, filename, path)
 
+    def put_atomic(self,
+         source_path: str,
+         destination_path: str) -> None:
+        upload_dir = f"{destination_path}.{uuid.uuid4()}.tmp"
+        try:
+            self.put(source_path, upload_dir)
+            self.mv(upload_dir, destination_path)
```
Comment (Totoketchup) [inline cluster_pack/filesystem.py:239]: I couldn't find the mv function in https://arrow.apache.org/docs/6.0/python/generated/pyarrow.fs.FileSystem.html, isn't it move ?

File: cluster_pack/uploader.py [hunk 7]
```diff
@@ -174,32 +185,49 @@ def _get_hash(spec_file: str) -> str:
...
-            into_to_upload = PexInfo.from_pex(zip_file)
-            if not force_upload and info_from_storage.code_hash == into_to_upload.code_hash:
-                _logger.info(f"skip upload of current {zip_file}"
-                             f" as it is already uploaded on {package_path}")
-                return
+        if not force_upload and _check_pex_version(resolved_fs, zip_file, package_path):
+            _logger.info(f"skip upload of current {zip_file}"
+                         f" as it is already uploaded on {package_path}")
+            return
+        elif resolved_fs.exists(package_path):
```
Comment (Totoketchup) [inline cluster_pack/uploader.py:196]: Was it this alreayd checked L191? If we are here I guess we know that the pex was already uploaded at this path but the pex versions are mismatching.

Plus I guess here you are removing the pex L198 because otherwise the put_atomic fails (just to be sure I got it) ?

File: cluster_pack/uploader.py [hunk 8]
```diff
@@ -174,32 +185,49 @@ def _get_hash(spec_file: str) -> str:
...
+            resolved_fs.rm(package_path)
 
     _logger.info(f"upload current {zip_file} to {package_path}")
 
     dir = os.path.dirname(package_path)
     if not resolved_fs.exists(dir):
         resolved_fs.mkdir(dir)
-    resolved_fs.put(zip_file, package_path)
+
+    if not _try_put_atomic(resolved_fs,
```
Comment (Totoketchup) [inline cluster_pack/uploader.py:206]: This function returning a boolean to then return an error seems not very clean IMO and it's not very easy to understand the logic behind it.

Couldn't we move the L210-212 to L228 ? If we receive a FileExistsError + verify_expected_output is False then we raise, otherwise it is ok, the pex matches

File: cluster_pack/uploader.py [hunk 9]
```diff
@@ -309,11 +340,29 @@ def _upload_env_from_venv(
         if not resolved_fs.exists(dir):
             resolved_fs.mkdir(dir)
         _logger.info(f'Uploading env at {local_package_path} to {package_path}')
-        resolved_fs.put(local_package_path, package_path)
+
+        if not _try_put_atomic(resolved_fs,
+                               local_package_path,
+                               package_path,
+                               lambda path: _is_archive_up_to_date(path, reqs, resolved_fs)):
```
Comment (Totoketchup) [inline cluster_pack/uploader.py:347]: Passing a function to a function is not super pythonic IMO + it seems the exact same function is passed on the second call you do to the _try_put_atomic function coudln't we integrate it to the _try_put_atomic function itself ?

File: cluster_pack/filesystem.py [hunk 10]
```diff
@@ -65,6 +68,23 @@ def _rm(self: Any, path: str, recursive: bool = False) -> None:
...
+    os.rename(path, new_path)
+
+
+def _s3_rename(self: Any, path: str, new_path: str) -> None:
+    if self.exists(new_path):
+        raise OSError(f"File exist at destination {new_path}")
+    self.fs.mv(path, new_path)
+
+
+def _hdfs_st_mode(self: Any, path: str) -> int:
```
Comment (Totoketchup) [inline cluster_pack/filesystem.py:83]: You added this for the tests ? (How could the `put` test have been working before then ? 🤔)

**Review by Totoketchup — CHANGES_REQUESTED** [review]
Some lil questions on the implementation that is not yet super clear to me

File: cluster_pack/filesystem.py (hunk 2 above)
Comment (j-rf) [reply cluster_pack/filesystem.py:238]: indeed, will move the try:

File: cluster_pack/filesystem.py (hunk 5 above)
Comment (j-rf) [reply cluster_pack/filesystem.py:217]: mv is calling rename in the base class

File: cluster_pack/filesystem.py (hunk 6 above)
Comment (j-rf) [reply cluster_pack/filesystem.py:239]: We are using the deprecated https://arrow.apache.org/docs/6.0/python/filesystems_deprecated.html

File: cluster_pack/filesystem.py (hunk 10 above)
Comment (j-rf) [reply cluster_pack/filesystem.py:83]: - rename was not used
- hdfs_st_mode is not actually used (except in tests I did locally with real hdfs)

File: cluster_pack/uploader.py (hunk 7 above)
Comment (j-rf) [reply cluster_pack/uploader.py:196]: Yes indeed. I will simplify this: after a review with Marc, the expected behavior is:
- fail if pex exists with the wrong version (unless force upload)
- remove existing only if force upload
BTW this will be the only code path modified, as other path were not subject to the race condition (see internal discussion on slack)

File: cluster_pack/uploader.py (hunk 8 above)
Comment (j-rf) [reply cluster_pack/uploader.py:206]: Ok indeed, I will simplify this

File: cluster_pack/uploader.py (hunk 9 above)
Comment (j-rf) [reply cluster_pack/uploader.py:347]: See comment above, will simplify, especially since now there is only one way to verify the package (as we no longer affect the code path from a venv, only from a pex)

File: tests/integration/test_skein_launcher.py (hunk 3 above)
Comment (j-rf) [reply tests/integration/test_skein_launcher.py:17]: Tests were failing, maybe could have been FAPR

File: cluster_pack/uploader.py (hunk 9 above)
Comment (j-rf) [reply cluster_pack/uploader.py:347]: To clarify: this commented change will disappear.

File: cluster_pack/uploader.py [hunk 11]
```diff
@@ -174,26 +179,33 @@ def _get_hash(spec_file: str) -> str:
...
-            into_to_upload = PexInfo.from_pex(zip_file)
-            if not force_upload and info_from_storage.code_hash == into_to_upload.code_hash:
-                _logger.info(f"skip upload of current {zip_file}"
-                             f" as it is already uploaded on {package_path}")
-                return
+        if force_upload:
+            _logger.info(f"forcing upload: removing current {package_path}")
+            resolved_fs.rm(package_path)
+        elif not _check_pex_version(resolved_fs, zip_file, package_path):
+            raise FileExistsError(f"Package {zip_file} already exists"
```
Comment (fhoering) [inline cluster_pack/uploader.py:190]: If the zip file exists with a different hash we need to override it. Otherwise we could just remove the file content checking and say that the pex file name is always unique when the content changes and in this case we just upload a new file name (removal not handled though)

File: cluster_pack/filesystem.py [hunk 12]
```diff
@@ -203,6 +230,19 @@ def put(self, filename: str, path: str, chunk: int = 2**16) -> None:
...
 
+    def put_atomic(self,
+         source_path: str,
+         destination_path: str) -> None:
+        tmp_upload_dir = f"{destination_path}.{uuid.uuid4()}.tmp"
+        self.put(source_path, tmp_upload_dir)
+        try:
+            self.mv(tmp_upload_dir, destination_path)
+        except OSError as err:
+            raise FileExistsError(f"File exist at destination {destination_path}", err)
```
Comment (fhoering) [inline cluster_pack/filesystem.py:241]: My understanding is that the existing put just overrides the file. It means `put` and `atomic_put` won't behave the same way.
My suggestion would be to delete before if the file exists, or we load the file and check its content before deleting.
In any case we can't do an atomic operation with HDFS, mv is not atomic either but our assumption is that delete and mv are significantly faster than the upload with put.
